junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/TempDirectory.java

```diff
@@ -177,22 +181,40 @@ private CleanupMode determineCleanupMode(TempDir tempDir) {
 		return cleanupMode == DEFAULT ? this.configuration.getDefaultTempDirCleanupMode() : cleanupMode;
 	}
 
-	private TempDirFactory getTempDirFactoryForField(Field field) {
+	@SuppressWarnings("deprecation")
+	private Scope getScope(ExtensionContext context) {
+		return context.getRoot().getStore(NAMESPACE).getOrComputeIfAbsent( //
+			Scope.class, //
+			__ -> new EnumConfigurationParameterConverter<>(Scope.class, "@TempDir scope") //
+					.get(TempDir.SCOPE_PROPERTY_NAME, context::getConfigurationParameter, Scope.PER_DECLARATION), //
+			Scope.class //
+		);
+	}
+
+	private TempDirFactory determineTempDirFactoryForField(Field field, Scope scope) {
 		TempDir tempDir = findAnnotation(field, TempDir.class).orElseThrow(
 			() -> new JUnitException("Field " + field + " must be annotated with @TempDir"));
-		return getTempDirFactory(tempDir);
+		return determineTempDirFactory(tempDir, scope);
 	}
 
-	private TempDirFactory getTempDirFactoryForParameter(ParameterContext parameterContext) {
+	private TempDirFactory determineTempDirFactoryForParameter(ParameterContext parameterContext, Scope scope) {
 		TempDir tempDir = parameterContext.findAnnotation(TempDir.class).orElseThrow(() -> new JUnitException(
 			"Parameter " + parameterContext.getParameter() + " must be annotated with @TempDir"));
-		return getTempDirFactory(tempDir);
+		return determineTempDirFactory(tempDir, scope);
 	}
 
-	private static TempDirFactory getTempDirFactory(TempDir tempDir) {
+	@SuppressWarnings("deprecation")
+	private TempDirFactory determineTempDirFactory(TempDir tempDir, Scope scope) {
 		Class<? extends TempDirFactory> factory = tempDir.factory();
+
+		if (factory != TempDirFactory.Standard.class && scope == Scope.PER_CONTEXT) {
+			throw new ExtensionConfigurationException("Custom @TempDir factory is not supported with "
+					+ TempDir.SCOPE_PROPERTY_NAME + "=" + Scope.PER_CONTEXT.toString().toLowerCase() + ". Use "
+					+ TempDir.DEFAULT_FACTORY_PROPERTY_NAME + " instead.");
+		}
+
 		return factory == TempDirFactory.Standard.class //
-				? TempDirFactory.Standard.INSTANCE
+				? this.configuration.getDefaultTempDirFactorySupplier().get()
```

Comment:
Unless I'm missing something... it appears that the logic here is broken.

If a user specifies their own "default" factory globally via the configuration parameter and then wants to actually use the `Standard` implementation for selected usage of `@TempDir`, then that is impossible with the current setup.

The proper way to handle default values for annotation attributes of type `Class` that allow a user to configure a strategy implementation is to have the attribute _default_ to the interface type instead of a concrete type.

In other words, `org.junit.jupiter.api.io.TempDir.factory()` should be declared as follows.

```java
Class<? extends TempDirFactory> factory() default TempDirFactory.class;
```

The logic for looking up what the user actually wants should then take that into account.

1. If the `factory` attribute is set to `TempDirFactory`, that means the user wants the default factory (either what's been configured globally or `Standard`).
2. If the `factory` attribute is set to anything else, that means the user wants to use that other thing.

Reply:
Actually, having `default TempDirFactory.class` or not having a default value at all should be the same, right?

Reply:
No. Not having a default makes it required.

Meaning... the user would always have to specify which factory type to use.

Having a default value for the attribute makes it an "optional" attribute -- which is what we want.

Reply:
Yep, mine was a silly question 🤦 😅

Reply:
FYI: I've encountered this type of use case several times over the years.

Spring's [`@ComponentScan(nameGenerator = ...)`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/annotation/ComponentScan.html#nameGenerator()) is a concrete example.

The Javadoc for that attribute describes the same sort of use case we have here with the `factory` for temporary directories.

> The default value of the [BeanNameGenerator](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/beans/factory/support/BeanNameGenerator.html) interface itself indicates that the scanner used to process this `@ComponentScan` annotation should use its inherited bean name generator, e.g. the default [AnnotationBeanNameGenerator](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/context/annotation/AnnotationBeanNameGenerator.html) or any custom instance supplied to the application context at bootstrap time.

Reply:
Done in 9eb31a55 and 29df048a.